src/mlpack/core/data/save_impl.hpp

```diff
@@ -38,220 +39,177 @@ bool Save(const std::string& filename,
-  }
-
-  // Try to save the file.
-  Log::Info << "Saving " << stringType << " to '" << filename << "'."
-      << std::endl;
-
-  // Transpose the matrix.
-  if (transpose)
-  {
-    arma::Mat<eT> tmp = trans(matrix);
-
-#ifdef ARMA_USE_HDF5
-    // We can't save with streams for HDF5.
-    const bool success = (saveType == FileType::HDF5Binary) ?
-        tmp.save(filename, ToArmaFileType(saveType)) :
-        tmp.save(stream, ToArmaFileType(saveType));
-#else
-    const bool success = tmp.save(stream, ToArmaFileType(saveType));
-#endif
-    if (!success)
-    {
-      Timer::Stop("saving_data");
-      if (fatal)
-        Log::Fatal << "Save to '" << filename << "' failed." << std::endl;
-      else
-        Log::Warn << "Save to '" << filename << "' failed." << std::endl;
-
-      return false;
-    }
-  }
-  else
-  {
-#ifdef ARMA_USE_HDF5
-    // We can't save with streams for HDF5.
-    const bool success = (saveType == FileType::HDF5Binary) ?
-        matrix.save(filename, ToArmaFileType(saveType)) :
-        matrix.save(stream, ToArmaFileType(saveType));
-#else
-    const bool success = matrix.save(stream, ToArmaFileType(saveType));
-#endif
-    if (!success)
-    {
-      Timer::Stop("saving_data");
-      if (fatal)
-        Log::Fatal << "Save to '" << filename << "' failed." << std::endl;
-      else
-        Log::Warn << "Save to '" << filename << "' failed." << std::endl;
-
-      return false;
-    }
-  }
+  MatrixOptions opts;
+  opts.Fatal() = fatal;
+  opts.NoTranspose() = !transpose;
+  opts.Format() = inputSaveType;
 
-  Timer::Stop("saving_data");
+  return Save(filename, matrix, opts);
+}
 
-  // Finally return success.
-  return true;
+template<typename MatType, typename DataOptionsType>
+bool Save(const std::string& filename,
+          const MatType& matrix,
+          const DataOptionsType& opts,
+          std::enable_if_t<IsArma<MatType>::value || IsSparseMat<MatType>::value>*)
+{
+  //! just use default copy ctor with = operator and make a copy.
+  DataOptionsType copyOpts(opts);
+  return Save(filename, matrix, copyOpts);
 }
 
-// Save a Sparse Matrix
-template<typename eT>
+/*
+ * Add this SFINAE in here because the compiler is so stupid that it is not
+ * able to distinguish between these two:
+ *
+ *  data::Save(filename, "model", *output);
+ *
+ *  and 
+ *
+ *  data::Save(filename, matrix, opts);
+ *  
+ * The second SFINAE is added because the compiler is bot able to see the
+ * difference between:
+ *
+ *  data::Save(filename, Row/Col, fatal);
+ *
+ *  and 
+ *
+ *  data::Save(filename, Row/Col, Opts);
+ *
+ * This SFINAE is temporary and must be removed after the integration of stage 3 or
+ * when the compiler becomes more intelligent.
+ */
+template<typename MatType, typename DataOptionsType>
 bool Save(const std::string& filename,
-          const arma::SpMat<eT>& matrix,
-          const bool fatal,
-          bool transpose)
+          const MatType& matrix,
+          DataOptionsType& opts,
```

Comment:
Actually, do we ever need to modify anything in `opts` when we are saving?  I don't think we do---in which case we could make it a `const` parameter, and we don't need the extra overload above that makes a copy of `opts`.

Reply:
Yes, we will need this, especially when we are trying to detect the file type. It was `const` initially, and then I did it like this.

Reply:
You're right!  I overlooked that.  Anyway having the non-const version is just fine.